### ISSUE TITLE
关于FTransUNet中的测试问题

### ISSUE THREAD

**Issue by yingzhige00** [sstary/SSRS#9]
在FTransUNet中，为什么在测试部分设置的Stride_Size只有32，导致测试部分过慢。
但是在写的test函数中默认stride=WINDOW_SIZE[0]，如果按照默认的设置，测试速度会变快，但是滑窗会变少，是否会影响最后的测试结果呢？

**Comment by sstary** [comment]
你好，因为训练过程中的测试比较频繁，为了更快地测试，在较大的数据集（LoveDA）上训练及测试时，可以使用较大的滑窗尺寸128，这种情况下测试出来的模型性能同样具有指示性，而滑窗32作为最终的测试，可以得到一个更高的指标。

**Comment by yingzhige00** [comment]
你好，对于模型的训练与测试现在都十分顺利，但是我看到您的论文中有关于注意力热力图的可视化图像，但是我实在是做不出来，您能分享下您的注意力图可视化部分的代码吗？十分感谢！

**Comment by yingzhige00** [comment]
太感谢了！！！！！

**Comment by yingzhige00** [comment]
大佬，模型调用处的代码我改在vitcross_seg_modeling.py的VisionTransformer的forward里面了，但是在x, y, attn_weights, features, cnn_x = self.transformer(x, y)  # (B, n_patch, hidden)处，他只输出了四个变量，但是却要五个变量来接受，他报错了。

**Comment by sstary** [comment]
当然会报错。这个cnn_x你根据你自己的需要从transformer里面传出来就好啦，后面的也是这样。

**Comment by yingzhige00** [comment]
原来如此

**Comment by yingzhige00** [comment]
大佬，再麻烦一下，我太菜了。
我在transformer中取出了Encoder返回的attn_weights列表的最后一个，它的形状是[1, 12, 256, 256]，而pred的形状是[1]， 他们在经过
`feature_grad = autograd.grad(pred, feature, allow_unused=True, retain_graph=True)[0]`
后，输出的是None，导致后面的无法进行，这是哪里有问题，该怎么做

**Comment by sstary** [comment]
你取的东西有问题，传出去的cnn_x就是正常数据流中的feature maps即可。比如我传出去的是：

    def forward(self, x, y):
        y = y.unsqueeze(1)
        if self.hybrid:
            x, y, features = self.hybrid_model(x, y)
        else:
            features = None
        cnn_x = x
        x = self.patch_embeddings(x)  # (B, hidden. n_patches^(1/2), n_patches^(1/2))
        y = self.patch_embeddingsd(y)
        x = x.flatten(2)
        x = x.transpose(-1, -2)  # (B, n_patches, hidden)
        y = y.flatten(2)
        y = y.transpose(-1, -2)
        
        embeddingsx = x + self.position_embeddings
        embeddingsx = self.dropout(embeddingsx)
        embeddingsy = y + self.position_embeddings
        embeddingsy = self.dropout(embeddingsy)
        # return embeddingsx, embeddingsy, features
        return embeddingsx, embeddingsy, features, cnn_x

这里的x，而不是attn_weights。这里改了后面class Transformer 的forward函数也得多一个参数。

**Comment by sstary** [comment]
![image](https://github.com/sstary/SSRS/assets/28852884/3154d375-4ff1-48a2-b5a5-2c2a3d70cb5e)

如图，这里shallow-feature的热力图为1*1024*16*16，pred由于取了类别和具体点，确实会成为一个值，也只能是一个值，因为一次只能绘制一个图。你先保证这些数据的维度和我这里一样看看。

**Comment by yingzhige00** [comment]
额，我现在哪怕直接采用上面的代码：
![image](https://github.com/sstary/SSRS/assets/133257451/13470085-2f86-422d-9645-83f7ae27130b)
![image](https://github.com/sstary/SSRS/assets/133257451/389b4014-8ba2-44db-9e42-e3ea08302d5c)
![image](https://github.com/sstary/SSRS/assets/133257451/dec80977-23e9-4221-ae82-b2b1f47052f2)
也会出现报错：
![image](https://github.com/sstary/SSRS/assets/133257451/cb7b51d1-1bab-4f38-809b-90545d71cd96)
不会是我的autograd有问题吧，这个我看原本代码是没有的，所以我用
`from torch import autograd`
导入了它

**Comment by sstary** [comment]
你好，你需要把test函数中这一行注释掉with torch.no_grad():。

**Comment by yingzhige00** [comment]
万分感谢，我已经跑成功了。还有就是在test_heatmap.py，貌似您忘了把BATCH_SIZE 设为1了，我在前面加上就可以正常跑了。

**Comment by sstary** [comment]
好的 ，已修改。

**Comment by sstary** [comment]
你好，都是可以的，这个绘制热力图的脚本符合所有的语义分割任务。

**Comment by wsyhhu** [comment]
非常感谢您的回复，还有一个问题，我该如何修改脚本指定目标类别。

**Comment by sstary** [comment]
该绘制代码是根据这一行
pred = logits[:, 3, 100, 65]
来确定的，实际绘制时，需要将batch size改为1，所以第一行为:，上面的3是类别，在utils中定义的，100和65和索引坐标。你可以对这一行进行调整并观察结果差异。

**Comment by wsyhhu** [comment]
好的，非常感谢。

**Comment by wsyhhu** [comment]
您好，我在运用到其他工作当中遇到了一个情况，heatmap1是全蓝或者全红的图片，而heatmap2，3是正常的热力图。请问是目标层设置不对吗？

**Comment by sstary** [comment]
你好，你需要检查你传入的特征，你可以对比一下，heatmap1和heatmap2的绘制代码是完全一致的，因此应该是你传入特征的问题。此外，这个绘制方式对输入目标应该没有要求，一般的特征图就能够根据回传梯度进行绘制。
